evidence/pool.go

```diff
@@ -75,59 +82,48 @@ func (evpool *Pool) PendingEvidence(maxNum uint32) []types.Evidence {
 	return evidence
 }
 
-// AllPendingEvidence returns all evidence ready to be proposed and committed.
-func (evpool *Pool) AllPendingEvidence() []types.Evidence {
-	evpool.removeExpiredPendingEvidence()
-	evidence, err := evpool.listEvidence(baseKeyPending, -1)
-	if err != nil {
-		evpool.logger.Error("Unable to retrieve pending evidence", "err", err)
-	}
-	return evidence
-}
-
 // Update uses the latest block & state to update any evidence that has been committed, to prune all expired evidence
 func (evpool *Pool) Update(block *types.Block, state sm.State) {
 	// sanity check
 	if state.LastBlockHeight != block.Height {
 		panic(fmt.Sprintf("Failed EvidencePool.Update sanity check: got state.Height=%d with block.Height=%d",
 			state.LastBlockHeight,
 			block.Height,
-		),
-		)
+		))
 	}
+	evpool.logger.Info("Updating evidence pool", "height", state.LastBlockHeight, "time", state.LastBlockTime)
 
 	// update the state
 	evpool.updateState(state)
 
 	// remove evidence from pending and mark committed
-	evpool.MarkEvidenceAsCommitted(block.Height, block.Evidence.Evidence)
+	evpool.markEvidenceAsCommitted(block.Height, block.Evidence.Evidence)
 
-	// prune pending, committed and potential evidence and polc's periodically
-	if block.Height%state.ConsensusParams.Evidence.MaxAgeNumBlocks == 0 {
-		evpool.logger.Debug("Pruning expired evidence")
-		// NOTE: As this is periodic, this implies that there may be some pending evidence in the
-		// db that have already expired. However, expired evidence will also be removed whenever
-		// PendingEvidence() is called ensuring that no expired evidence is proposed.
-		evpool.removeExpiredPendingEvidence()
+	// prune pending evidence when it has expired. This also updates when the next evidence will expire
+	if evpool.evidenceSize > 0 && state.LastBlockHeight > evpool.pruningHeight &&
+		state.LastBlockTime.After(evpool.pruningTime) {
+		evpool.pruningHeight, evpool.pruningTime = evpool.removeExpiredPendingEvidence()
 	}
 }
 
 // AddEvidence checks the evidence is valid and adds it to the pool.
 func (evpool *Pool) AddEvidence(ev types.Evidence) error {
 	evpool.logger.Debug("Attempting to add evidence", "ev", ev)
 
-	if evpool.Has(ev) {
-		return nil
+	// We have already verified this piece of evidence - no need to do it again
+	if evpool.isPending(ev) {
+		return fmt.Errorf("evidence already verified and added, %s", ev.String())
```

Comment:
```suggestion
		return fmt.Errorf("evidence %v already verified and added", ev)
```

Reply:
Do we even need this evidence.String() function?

Reply:
also I prefer putting the `%v` part at the end because it destroys the sentence

Reply:
%v will automatically call `String()` func if defined on Evidence